### ISSUE TITLE
Help is broken in 4.25 M2

### ISSUE THREAD

**Issue by iloveeclipse** [eclipse-equinox/equinox#88]
Download latest SDK.
Open Help -> Help contents.
Kaboom.

![image](https://user-images.githubusercontent.com/964108/180980196-e2bb6956-9827-4cb2-86a1-e0e84d0836c2.png)

Same issue as https://github.com/eclipse-platform/eclipse.platform.ua/issues/42 , unfortunately I haven't realized that this is **not a tests only issue**, because in my own SDK IDE I've installed  org.apache.commons.fileupload bundle vis other dependencies.

There are errors in the log:

```
eclipse.buildId=4.25.0.I20220725-1800
java.version=17.0.3
java.vendor=Eclipse Adoptium
BootLoader constants: OS=linux, ARCH=x86_64, WS=gtk, NL=en_US
Command-line arguments:  -os linux -ws gtk -arch x86_64

org.eclipse.equinox.http.servlet
Error
Tue Jul 26 11:51:40 CEST 2022
org.eclipse.equinox.http.servlet.internal.multipart.MultipartSupportFactory: Provider org.eclipse.equinox.http.servlet.internal.multipart.MultipartSupportFactoryImpl could not be instantiated

java.util.ServiceConfigurationError: org.eclipse.equinox.http.servlet.internal.multipart.MultipartSupportFactory: Provider org.eclipse.equinox.http.servlet.internal.multipart.MultipartSupportFactoryImpl could not be instantiated
	at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:586)
	at java.base/java.util.ServiceLoader$ProviderImpl.newInstance(ServiceLoader.java:813)
	at java.base/java.util.ServiceLoader$ProviderImpl.get(ServiceLoader.java:729)
	at java.base/java.util.ServiceLoader$3.next(ServiceLoader.java:1403)
	at org.eclipse.equinox.http.servlet.internal.registration.ServletRegistration.<clinit>(ServletRegistration.java:40)
	at org.eclipse.equinox.http.servlet.internal.context.ContextController.doAddServletRegistration(ContextController.java:478)
	at org.eclipse.equinox.http.servlet.internal.context.ContextController.addServletRegistration(ContextController.java:435)
	at org.eclipse.equinox.http.servlet.internal.customizer.ContextServletTrackerCustomizer.addingService(ContextServletTrackerCustomizer.java:81)
	at org.eclipse.equinox.http.servlet.internal.customizer.ContextServletTrackerCustomizer.addingService(ContextServletTrackerCustomizer.java:1)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:943)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:1)
	at org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256)
	at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:229)
	at org.osgi.util.tracker.ServiceTracker$Tracked.serviceChanged(ServiceTracker.java:903)
	at org.eclipse.osgi.internal.serviceregistry.FilteredServiceListener.serviceChanged(FilteredServiceListener.java:123)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:961)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:234)
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:151)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEventPrivileged(ServiceRegistry.java:937)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEvent(ServiceRegistry.java:874)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.register(ServiceRegistrationImpl.java:141)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.registerService(ServiceRegistry.java:262)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.registerService(BundleContextImpl.java:500)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.registerService(BundleContextImpl.java:519)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.registerService(BundleContextImpl.java:1047)
	at org.eclipse.equinox.http.servlet.internal.HttpServiceRuntimeImpl.registerHttpServiceServlet(HttpServiceRuntimeImpl.java:891)
	at org.eclipse.equinox.http.servlet.internal.HttpServiceImpl$3.run(HttpServiceImpl.java:137)
	at org.eclipse.equinox.http.servlet.internal.HttpServiceImpl$3.run(HttpServiceImpl.java:1)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:569)
	at org.eclipse.equinox.http.servlet.internal.HttpServiceImpl.registerServlet(HttpServiceImpl.java:134)
	at org.eclipse.equinox.http.registry.internal.HttpRegistryManager.registerServlet(HttpRegistryManager.java:261)
	at org.eclipse.equinox.http.registry.internal.HttpRegistryManager.addServletContribution(HttpRegistryManager.java:143)
	at org.eclipse.equinox.http.registry.internal.ServletManager.added(ServletManager.java:137)
	at org.eclipse.equinox.http.registry.internal.ExtensionPointTracker.open(ExtensionPointTracker.java:80)
	at org.eclipse.equinox.http.registry.internal.ServletManager.start(ServletManager.java:65)
	at org.eclipse.equinox.http.registry.internal.HttpRegistryManager.start(HttpRegistryManager.java:109)
	at org.eclipse.equinox.http.registry.internal.HttpServiceTracker.addingService(HttpServiceTracker.java:61)
	at org.eclipse.equinox.http.registry.internal.HttpServiceTracker.addingService(HttpServiceTracker.java:1)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:943)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:1)
	at org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256)
	at org.osgi.util.tracker.AbstractTracked.trackInitial(AbstractTracked.java:183)
	at org.osgi.util.tracker.ServiceTracker.open(ServiceTracker.java:321)
	at org.osgi.util.tracker.ServiceTracker.open(ServiceTracker.java:264)
	at org.eclipse.equinox.http.registry.internal.HttpServiceTracker.open(HttpServiceTracker.java:45)
	at org.eclipse.equinox.http.registry.internal.Activator.addingService(Activator.java:63)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:943)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:1)
	at org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256)
	at org.osgi.util.tracker.AbstractTracked.trackInitial(AbstractTracked.java:183)
	at org.osgi.util.tracker.ServiceTracker.open(ServiceTracker.java:321)
	at org.osgi.util.tracker.ServiceTracker.open(ServiceTracker.java:264)
	at org.eclipse.equinox.http.registry.internal.Activator.start(Activator.java:41)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$2.run(BundleContextImpl.java:818)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$2.run(BundleContextImpl.java:1)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:569)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.startActivator(BundleContextImpl.java:810)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.start(BundleContextImpl.java:767)
	at org.eclipse.osgi.internal.framework.EquinoxBundle.startWorker0(EquinoxBundle.java:1032)
	at org.eclipse.osgi.internal.framework.EquinoxBundle$EquinoxModule.startWorker(EquinoxBundle.java:371)
	at org.eclipse.osgi.container.Module.doStart(Module.java:605)
	at org.eclipse.osgi.container.Module.start(Module.java:468)
	at org.eclipse.osgi.internal.framework.EquinoxBundle.start(EquinoxBundle.java:445)
	at org.eclipse.help.internal.server.JettyHelpServer.checkBundle(JettyHelpServer.java:132)
	at org.eclipse.help.internal.server.JettyHelpServer.start(JettyHelpServer.java:119)
	at org.eclipse.help.internal.server.WebappManager.start(WebappManager.java:41)
	at org.eclipse.help.internal.base.BaseHelpSystem.ensureWebappRunning(BaseHelpSystem.java:186)
	at org.eclipse.help.internal.base.HelpDisplay.displayHelpURL(HelpDisplay.java:175)
	at org.eclipse.help.internal.base.HelpDisplay.displayHelp(HelpDisplay.java:72)
	at org.eclipse.help.ui.internal.DefaultHelpUI.displayHelp(DefaultHelpUI.java:143)
	at org.eclipse.ui.internal.help.WorkbenchHelpSystem.displayHelp(WorkbenchHelpSystem.java:706)
	at org.eclipse.ui.internal.actions.HelpContentsAction.lambda$0(HelpContentsAction.java:81)
	at org.eclipse.swt.custom.BusyIndicator.showWhile(BusyIndicator.java:74)
	at org.eclipse.ui.internal.actions.HelpContentsAction.run(HelpContentsAction.java:81)
	at org.eclipse.jface.action.Action.runWithEvent(Action.java:474)
	at org.eclipse.jface.action.ActionContributionItem.handleWidgetSelection(ActionContributionItem.java:580)
	at org.eclipse.jface.action.ActionContributionItem.lambda$4(ActionContributionItem.java:414)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5794)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1529)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:5025)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4477)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1155)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:643)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:550)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:156)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:402)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1440)
Caused by: java.lang.NoClassDefFoundError: org/apache/commons/fileupload/FileUploadException
	at org.eclipse.equinox.http.servlet.internal.multipart.MultipartSupportFactoryImpl.<clinit>(MultipartSupportFactoryImpl.java:23)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480)
	at java.base/java.util.ServiceLoader$ProviderImpl.newInstance(ServiceLoader.java:789)
	... 102 more
Caused by: java.lang.ClassNotFoundException: org.apache.commons.fileupload.FileUploadException cannot be found by org.eclipse.equinox.http.servlet_1.7.300.v20220720-2012
	at org.eclipse.osgi.internal.loader.BundleLoader.generateException(BundleLoader.java:541)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass0(BundleLoader.java:536)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:416)
	at org.eclipse.osgi.internal.loader.ModuleClassLoader.loadClass(ModuleClassLoader.java:168)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	... 109 more
```

**Comment by laeubi** [comment]
`org.eclipse.equinox.http.servlet` should really import `org.apache.commons.fileupload` package as a hard requirement then....

**Comment by HannesWell** [comment]
> `org.eclipse.equinox.http.servlet` should really import `org.apache.commons.fileupload` package as a hard requirement then....

Yep, I think so as well.
And as noted in https://github.com/eclipse-platform/eclipse.platform.ua/issues/42#issuecomment-1193275873, probably some more sub-packages.

From my knowlegde you can only mark a dependency as optional If you wrap all instanciations of corresponding classes in try-catch-blocks that catch the `java.lang.ClassNotFoundException`.

**Comment by laeubi** [comment]
> From my knowlegde you can only mark a dependency as optional If you wrap all instanciations of corresponding classes in try-catch-blocks that catch the `java.lang.ClassNotFoundException`.

It all depends a bit how gracefully you want to handle such cases, in general I found it not worth the effort at all, if something is really "optional" one better extract it into an own module.

**Comment by iloveeclipse** [comment]
Note: this is **not** a problem of not existing optional package!

- SDK build I20220719-1800 **doesn't contain** `org.apache.commons.fileupload` bundle but **works fine**!
- SDK build I20220721-0550 however is broken.

So it looks like either equinox or something else changed in the way the `org.eclipse.equinox.http.servlet.internal.multipart.MultipartSupportFactoryImpl` class is loaded on help startup.

**Comment by iloveeclipse** [comment]
See fix in #89. The root cause **was** the cleanup from https://github.com/eclipse-equinox/equinox/pull/45, as I've assumed in https://github.com/eclipse-platform/eclipse.platform.ua/issues/42.

**Comment by iloveeclipse** [comment]
Verified with I20220726-1800 build, help works again.
